internal/provider/tar_archiver.go

```diff
@@ -0,0 +1,273 @@
+}
+
+func NewTarArchiver(filepath string, compression TarCompressionType) Archiver {
+	return &TarArchiver{
+		filepath:    filepath,
+		compression: compression,
+	}
+}
+
+func (a *TarArchiver) ArchiveContent(content []byte, infilename string) error {
+	if err := a.open(); err != nil {
+		return err
+	}
+	defer a.close()
+
+	return a.addContent(content, &tar.Header{
+		Name:    infilename,
+		Size:    int64(len(content)),
+		ModTime: time.Time{},
+	})
+}
+
+func (a *TarArchiver) ArchiveFile(infilename string) error {
+	fi, err := assertValidFile(infilename)
+	if err != nil {
+		return err
+	}
+
+	if err := a.open(); err != nil {
+		return err
+	}
+	defer a.close()
+
+	if err := a.addFile(infilename, filepath.ToSlash(fi.Name()), time.Time{}); err != nil {
+		return err
+	}
+
+	return err
+}
+
+func (a *TarArchiver) ArchiveDir(indirname string, opts ArchiveDirOpts) error {
+	_, err := assertValidDir(indirname)
+	if err != nil {
+		return err
+	}
+
+	// ensure exclusions are OS compatible paths
+	for i := range opts.Excludes {
+		opts.Excludes[i] = filepath.FromSlash(opts.Excludes[i])
+	}
+
+	if err := a.open(); err != nil {
+		return err
+	}
+	defer a.close()
+
+	return filepath.Walk(indirname, a.createWalkFunc("", indirname, opts))
+}
+
+func (a *TarArchiver) createWalkFunc(basePath string, indirname string, opts ArchiveDirOpts) func(path string, info os.FileInfo, err error) error {
+	return func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("error encountered during file walk: %s", err)
+		}
+
+		relname, err := filepath.Rel(indirname, path)
+		if err != nil {
+			return fmt.Errorf("error relativizing file for archival: %s", err)
+		}
+
+		archivePath := filepath.Join(basePath, relname)
+
+		isMatch := checkMatch(archivePath, opts.Excludes)
+
+		if info.IsDir() {
+			if isMatch {
+				return filepath.SkipDir
+			}
+			return nil
+		}
+
+		if isMatch {
+			return nil
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if info.Mode()&os.ModeSymlink == os.ModeSymlink {
+			if !opts.ExcludeSymlinkDirectories {
+				path, err = filepath.EvalSymlinks(path)
+				if err != nil {
+					return err
+				}
+
+				info, err = os.Stat(path)
+				if err != nil {
+					return err
+				}
+
+				if info.IsDir() {
+					return filepath.Walk(path, a.createWalkFunc(archivePath, path, opts))
+				}
+			}
+		}
+
+		return a.addFile(path, archivePath, info.ModTime())
+	}
+}
+
+func (a *TarArchiver) ArchiveMultiple(content map[string][]byte) error {
+	if err := a.open(); err != nil {
+		return err
+	}
+	defer a.close()
+
+	// Ensure files are processed in the same order so hashes don't change
+	keys := make([]string, len(content))
+	i := 0
+	for k := range content {
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+
+	for _, filename := range keys {
+		header := &tar.Header{
+			Name:    filepath.ToSlash(filename),
+			Size:    int64(len(content[filename])),
+			ModTime: time.Now().Round(time.Second),
```

Comment:
I guess the change makes sense, however a bug should be covered by an additional test to prevent that the bug may appears in the future again.

Hanging on that a bit

Reply:
I was able to reproduce it

Reply:
https://registry.terraform.io/providers/jkroepke/archive/2.4.2+tar.gz-20240510134300